fix Piecewise assumptions

smichr opened the pull request:
fixes #10258

smichr commented on sympy/functions/elementary/tests/test_piecewise.py:
```diff
@@ -495,3 +500,19 @@ def test_S_srepr_is_identity():
+
+
+def test_issue_10258():
+    assert Piecewise((0, x < 1), (1, True)).is_zero is False
```
note to self: perhaps better to return None (if possible) keep old code but instead of return when_multiple, return None (and keep the Expr header test).

asmeurer commented:
Yeah, I've wondered about that commutative thing before. Do tests break if you change that (I vaguely remember something depends on it)? 

Maybe it would be helpful at the very least to fix that after https://github.com/sympy/sympy/issues/6494, so that people can easily write `Function('f', commutative=False)`.

asmeurer commented on sympy/functions/elementary/tests/test_piecewise.py:
```diff
@@ -495,3 +495,20 @@ def test_S_srepr_is_identity():
+    assert Piecewise((1, x < 1), (2, True)).is_nonzero
+    assert Piecewise((0, x < 1), (oo, True)).is_finite is None
+    assert Piecewise((0, x < 1), (1, True)).is_finite
+    assert Piecewise(([1], x < 1)).is_finite is None
```
What is this? You shouldn't be able to put lists in SymPy objects.

smichr replied:
I was surprised, too, but this is necessary as long as anything in SymPy can emit a list, e.g. solve. A Piecewise is an appropriate container for handling any objects (Basic or otherwise) whose selection depends on some condition. The only way to represent a solution to an equation with piecewise conditions is with a Piecewise of lists (or dicts).

For a univariate Piecewise we can give this:

```
>> solve(Piecewise((x**2-4,x>-4),(x+6,True)), x)
[-6, -2, 2]
```
But for a multivariate one we need to give a piecewise result:
```
>> solve(Piecewise((x**2-4,y>0),(x+6,True)),x)
Piecewise(([-2, 2], y > 0), (-6, True))
```

asmeurer replied:
This shouldn't be allowed. The result should be a list of Piecewise, not the other way around. No SymPy object should have args that is a list, since list isn't Basic (and it's mutable, which is double bad).

smichr replied:
I was puzzled by this when I encountered it. What object will you return so that when you substitute y=1 you get the list [-2. 2]? The syntax is solve(function) -> `list of solutions` not `[list of solutions]`, so `[-2, 2]` not `[[-2, 2]]` Piecewise allows us to say "this list under this condition otherwise this list."

smichr replied:
We could just hack around this by making Piecewise return its solve results in a Tuple when necessary and documenting accordingly.

asmeurer replied:
Why not `[Piecewise((-2, y > 0), (-6, True)), Piecewise((2, y > 0), (-6, True))]`? That will give `[-2, 2]` if you substitute y=1 into each solution.

smichr replied:
So we yield a number of items equal to the longest solution and accept that there will be repeats?

asmeurer replied:
There are only repeats for y < 0. I see this as similar to `solve(x**2 - a, x)` -> `[sqrt(a), -sqrt(a)]`, which contains the same solution twice for a=0.

smichr replied:
Leave a comment

> I see this as similar to

hmm. That makes sense -- when not given enough information to filter the results, duplicates may be returned.

smichr replied:
I removed the mods to Function.is_commutative (sans typo). I vaguely recall that the case with a non-Expr arg will now fail because assumptions are worked out based on commutativity being False instead of None.  We'll see.

asmeurer replied:
You could be right about that. Might end up being a harder thing to fix than expected.

asmeurer commented on sympy/functions/elementary/piecewise.py:
```diff
@@ -462,20 +465,22 @@ def _eval_subs(self, old, new):
     def _eval_transpose(self):
         return self.func(*[(e.transpose(), c) for e, c in self.args])
 
-    def _eval_template_is_attr(self, is_attr, when_multiple=None):
```
How does removing this flag fix the issue?

smichr replied:
`when_multiple` was used to make it emit something other than None when a multivalued expressions existed, e.g. `Piecewise((0, x< 0), (1, True)).is_zero` -> False but it should give None: the conditions do not allow us to say definitievely that the piecewise is zero. When we have `Piecewise((2, x< 0), (1, True)).is_zero` *then* we can say False.

asmeurer replied:
Is it doing anything smart enough to check if `x.is_negative`, or is that outside the scope of this PR?

smichr replied:
One would only inspect the condition if the expression were a function of a common variable. And no, that is not included and cannot be until I get to the point of having the robust condition checking in place -- it is currently very easy to trip up the condition checking. Once that is in place, any condition sharing a single variable with the expression could be solved and substituted into the expression when doing the checking so Piecewise arg

```
(a + b - 3, a + b < 2)` -> solve cond for a and subst using positive symbol `pos` to indicate that we are using "<"
(2 - b - pos) + b - 3 ->
-1 - pos -> a negative
```
could be inferred.

smichr commented:
Leave a comment

> Do tests break if you change that

I modified the function to return None if there is a mix of commutative and noncommutative args. Nothing broke (which is not too surprising). What did you have in mind to change, returning None if any arg is noncommutative?

asmeurer commented:
I went ahead and implemented undefined function assumptions https://github.com/sympy/sympy/pull/12945. So I think with that PR we should definitely make `Function('f')(x).is_commutative` unrelated to `x.is_commutative`, since you can easily set `Function('f', commutative=False)`.

asmeurer commented:
Can the is_commutative stuff be moved to a separate PR?

smichr commented:
> moved to a separate PR?

yes

smichr committed:
10258: fix Piecewise assumptions

When there is a disagreement between expressions in the
Piecewise, None should be returned as no answer can be
given regarding the assumption.
